### DISCUSSION TITLE
Run tasks for duration of client connection (i.e. start/stop on connect/disconnect)

### DISCUSSION THREAD

**Discussion by TimChild** [reflex-dev/reflex#3105] [Q&A]
I'd like to run some tasks in the background that can periodically push updates to the user, but I'm not sure how best to handle the lifetime of the task. 

To give a bit more context, here is a minimal example of the sort of thing I'm trying to do.
```
class SomeState(rx.State):
    some_value: int = 0
    
    @rx.background
    async def long_running_task(self):
        r = get_redis()
        async with r.pubsub(ignore_subscribe_messages=True) as pubsub:
            await pubsub.subscribe("some_client_specific_channel")
            async for message in pubsub.listen():
                if message["type"] == "message":
                    if message['data'] == 'STOP':
                        break
                    else:
                        self.some_value = int(message["data"])
```

@masenf Gave a nice suggestion a while ago in #1857 that was similar. This is the relevant part:
```
app = rx.App()
app.compile()


# Handle websocket disconnect events to avoid memory leaks when sessions are closed
orig_disconnect = app.event_namespace.on_disconnect

def disconnect_handler(sid):
    orig_disconnect(sid)

    clientToken, token = tokens_by_sid.get(sid, (None, None))
    print(f"Disconnect event received for {sid}. Removing {token} from shared {clientToken}")

    shared_sessions_by_token.get(clientToken, set()).discard(token)
    tokens_by_sid.pop(sid, None)

app.event_namespace.on_disconnect = disconnect_handler
```

Where something similar could be done for `on_connect`. 

This is close to what I am looking for in that the events trigger at client connect/disconnect. I'm just not sure how I can start a background task from the `on_connect` equivalent. In my case, I'd use the `on_disconnect` to send `STOP` to  "some_client_specific_channel". 

There is also this #1970, which is also similar in nature. In fact, I am currently using a recommendation there to use `rx.moment` to periodically start the long running background task with a timeout that will prevent it running forever if the user closes the tab. 
In this case, to prevent two tasks running at once per user, I just send a `STOP` to the channel immediately before beginning to listen again. Overall, this is a bit messy though, and I'd prefer to just use the actual connect/disconnect events directly.

Or is there a much simpler approach I am overlooking entirely?

**Comment by benedikt-bartscher** [comment]
afaik there is currently no simpler approach. I wanted to simplify the connection state here: https://github.com/reflex-dev/reflex/pull/2671  
I also had some ideas to implement a generalized one-task-per-session decorator.

With https://github.com/reflex-dev/reflex/issues/2771 we could even extend this idea to a one-task-per-user decorator.

cc @abulvenz

**Reply by TimChild** [reply]
Well, I'm glad I wasn't missing some thing obvious at least! 

Yea, it would be nice to combine with #2771
